### PR TITLE
Update pairing logic to allow clinical normal and tumor research samples

### PR DESCRIPTION
**Pull request by alexiswl**
* Don't skip immediately if cannot find matching tumor or normal.
* First check if tumor is in run and workflow group is 'research'.
  * Then check if there is a matching clinical normal if there are no matching research normals
* Second check if normal is in run and workflow is clinical
  * Then check  if there is a matching research tumor if there are no matching clinical tumors.

* Get a unique list of jobs in the event that a clinical normal and research tumor are on the same run.

* Added filter for library ids in get_workflows_by_subject_id. We don't want to wait for libraries that might not be involved in this pairing.

* Added tests for running 'mixed types' between research and clinical

### CODE REVIEW

**Comment by victorskl** [comment]
Thanks Alexis. I will review soon.

**Comment by victorskl** [comment]
[Portal Release - Ocicat: Support mixed clinical/research WGS pairing](https://trello.com/c/SpYWV9q1/1476-portal-release-ocicat-support-mixed-clinical-research-wgs-pairing)

File: data_processors/pipeline/orchestration/tests/test_tumor_normal_step.py [hunk 1]
```diff
@@ -201,6 +299,45 @@ def test_prepare_tumor_normal_jobs_issue_475_skip(self):
...
+
+        mock_lbr_normal: LibraryRun = LibraryRunFactory()
+        mock_lbr_tumor: LibraryRun = TumorLibraryRunFactory()
+
+        meta_list = LabMetadata.objects.all()
+
+        job_list, subjects, submitting_subjects = tumor_normal_step.prepare_tumor_normal_jobs(meta_list)
+
+        logger.info(f"job_list: {json.dumps(job_list)}")
+
```
Comment (victorskl) [inline data_processors/pipeline/orchestration/tests/test_tumor_normal_step.py:340]: Great stuff! Thanks for unit testing. 👍

**Review by victorskl — APPROVED** [review]
LGTM! Minor comment to attend.

File: data_processors/pipeline/orchestration/tests/test_tumor_normal_step.py (hunk 1 above)
Comment (victorskl) [reply data_processors/pipeline/orchestration/tests/test_tumor_normal_step.py:340]: Ah. One thing. Could you add assertion pls. Like, `len(job_list) == 0` for example.

File: data_processors/pipeline/orchestration/tumor_normal_step.py [hunk 2]
```diff
@@ -157,6 +150,40 @@ def prepare_tumor_normal_jobs(meta_list: List[LabMetadata]) -> (List, List, List
...
+            if tumor_in_run and len(subject_normal_libraries_stripped) == 0 and workflow == LabMetadataWorkflow.RESEARCH.value:
+                # Re collect clinical normal libraries
+                subject_normal_libraries: List[str] = metadata_srv.get_wgs_normal_libraries_by_subject(
+                    subject_id=subject,
+                    meta_workflow=str(LabMetadataWorkflow.CLINICAL.value)
+                )
+                subject_normal_libraries_stripped = _mint_libraries(subject_normal_libraries)
+                if len(subject_normal_libraries_stripped) > 0:
+                    logger.info(f"Pairing clinical tumors(s) in run '{', '.join(subject_tumor_libraries_stripped)}' "
+                                f"with research normal(s) '{', '.join(subject_normal_libraries_stripped)}'")
```
[the remark this answers is not in the document]
Comment (alexiswl) [reply data_processors/pipeline/orchestration/tumor_normal_step.py]: Resolved in f59e06500b0f225a4a86a55e596e8666d204f0c9

File: data_processors/pipeline/orchestration/tumor_normal_step.py [hunk 3]
```diff
@@ -157,6 +150,40 @@ def prepare_tumor_normal_jobs(meta_list: List[LabMetadata]) -> (List, List, List
...
+            if normal_in_run and len(subject_tumor_libraries_stripped) == 0 and workflow == LabMetadataWorkflow.CLINICAL.value:
+                # Re collect research tumor libraries
+                subject_tumor_libraries: List[str] = metadata_srv.get_wgs_tumor_libraries_by_subject(
+                    subject_id=subject,
+                    meta_workflow=str(LabMetadataWorkflow.RESEARCH.value)
+                )
+                subject_tumor_libraries_stripped = _mint_libraries(subject_tumor_libraries)
+                if len(subject_tumor_libraries_stripped) > 0:
+                    logger.info(f"Pairing clinical normal(s) in run '{', '.join(subject_normal_libraries_stripped)}' "
+                                f"with research tumour(s) '{', '.join(subject_tumor_libraries_stripped)}'")
```
[the remark this answers is not in the document]
Comment (alexiswl) [reply data_processors/pipeline/orchestration/tumor_normal_step.py]: Resolved in f59e06500b0f225a4a86a55e596e8666d204f0c9

File: data_processors/pipeline/orchestration/tests/test_tumor_normal_step.py (hunk 1 above)
Comment (alexiswl) [reply data_processors/pipeline/orchestration/tests/test_tumor_normal_step.py:341]: Resolved in f59e06500b0f225a4a86a55e596e8666d204f0c9

**Review by victorskl — APPROVED** [review]
Good to go this, Alexis. Thanks

**Comment by victorskl** [comment]
Updates made in Portal 2.2 (Ocicat) release to the original TN pairing from PR https://github.com/umccr/data-portal-apis/pull/262

**Pull request merged**